### ISSUE TITLE
Features which Interval / IntervalIndex should probably have

### ISSUE THREAD

**Issue by zfrenchee** [pandas-dev/pandas#19480]
Here's a list of possible features which should be added to the Interval and IntervalIndex types. Some of them might already exist, in which case, please excuse my mistake. **Also note**, I'm not asking for these, just listing some ideas I had which might be useful for others.  

- **Closest**
```
>>> Interval(3, 4).closest(IntervalIndex.from_tuples([(1, 2), (5, 9)]), how="min|mean|max")
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).closest(IntervalIndex.from_tuples([(1, 4), (2, 5)]))
```
Big question: what if there are two or more identically distant intervals? 

- **Complement**
```
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).complement(lower_bound=np.neginf, upper_bound=np.inf)
IntervalIndex.from_tuples([(np.neginf, 1), (2, 5), (9, np.inf)])
```

- **intersection**
```
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).intersect(IntervalIndex.from_tuples([(3, 4), (6, 7)]))
IntervalIndex.from_tuples([(6, 7)])
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).intersect(Interval(6, 7))
???
```

- **union**
```
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).intersect(IntervalIndex.from_tuples([(3, 4), (6, 7)]))
IntervalIndex.from_tuples([(1, 2), (3, 4), (5, 9)])
```

- **subtract|difference**
```
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]) - IntervalIndex.from_tuples([(3, 4), (6, 7)])
IntervalIndex.from_tuples([(1, 2), (5, 6), (7, 9)])
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]) - Interval(6, 7)
IntervalIndex.from_tuples([(1, 2), (5, 6), (7, 9)])
>>> Interval(6, 7) - IntervalIndex.from_tuples([(1, 2), (5, 9)])
???
```

- **sort**
 (I'm guessing this is probably already implicitly implemented. Does it work for multi-index in which one of the levels is Intervalindex?)

- **shift**
``
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).shift(2)
IntervalIndex.from_tuples([(3, 4), (7, 11)])
```
(Could be really useful for datetimes?)

- **slop|grow|window|better name**
```
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).slop(1)
IntervalIndex.from_tuples([(0, 3), (4, 10)])
```

**Comment by jschendel** [comment]
Note that `union`/`intersection`/`difference`/`shift` will likely need new names, as they already have more generic uses (e.g. `IntervalIndex.union` is already implemented with set theoretic behavior in terms of individual elements).

Some more ideas, all of which should return booleans, mostly taken from [postgres range types](https://www.postgresql.org/docs/current/static/functions-range.html):
- `is_adjacent_to`
- `strictly_left_of`
- `strictly_right_of`
- `is_empty`

(see the postgres docs for examples)

Another postgres function that could be nice is `range_merge`, which we'd almost certainly want to rename to something like `combine`.  It gives the smallest interval containing two given intervals, e.g. `Interval(0, 1).combine(Interval(3, 4))` would return `Interval(0, 4)`.

**Comment by tsabsch** [comment]
I was currently working with an IntervalIndex and needed to get the interval ranges. Maybe this would be useful for others as well?
```
>>> IntervalIndex.from_tuples([(1, 2), (5, 9)]).ranges()
array([1, 4])
```

**Comment by jschendel** [comment]
@tsabsch : By `ranges`, you're referring to the length of each interval in the `IntervalIndex`, right?  

If so, note that this was implemented as the `length` property in #18805, and will be present in 0.23.0:
```python
In [2]: pd.__version__
Out[2]: '0.23.0.dev0+230.gf391cbf'

In [3]: pd.IntervalIndex.from_tuples([(1, 2), (5, 9)]).length
Out[3]: Int64Index([1, 4], dtype='int64')
```

**Comment by tsabsch** [comment]
@jschendel Yes, this is exactly what I meant. Thank you for pointing it out. I wasn't aware that it was already implemented for the next release. I'm on pandas 0.22.0.

**Comment by jschendel** [comment]
@tsabsch : No worries.  Glad to hear `length` will be useful!

**Comment by mezzode** [comment]
It might also be nice to be able to check if an `Interval` is `in` another `Interval`.
Also, this issue could use the `Interval` label others for discoverability.
